src/Authenticators/TOTPAuthenticator.php

```diff
@@ -26,47 +28,54 @@ class TOTPAuthenticator extends BootstrapMFAAuthenticator
      * a specific TOTP authenticator
      */
     private static $algorithm = 'sha1';
-
-
+    
     /**
-     * @param string $link
-     * @return \SilverStripe\Security\MemberAuthenticator\LoginHandler|static
+     * Get configured algorithm for TOTP Authenticator
+     *
+     * Must be one of: "sha1", "sha256", "sha512"
+     * If not specified or invalid, default to "sha1"
+     * @return string
      */
-    public function getLoginHandler($link)
+    public static function get_algorithm()
     {
-        return TOTPLoginHandler::create($link, $this);
+        $algorithm = self::config()->get('algorithm');
+
+        return in_array(strtolower($algorithm), ['sha1', 'sha256', 'sha512'])
+            ? $algorithm
+            : 'sha1';
     }
 
     /**
      * @param $data
      * @param HTTPRequest $request
+     * @param $token
      * @param ValidationResult $result
      * @return bool|null|Member
      * @throws \Exception
      */
-    public function validateTOTP($data, $request, &$result)
+    public function verifyMFA($data, $request, $token, &$result)
     {
         $memberID = $request->getSession()->get(BootstrapMFAAuthenticator::SESSION_KEY . '.MemberID');
 
         // First, let's see if we know the member
-        /** @var Member $member */
+        /** @var Member|null $member */
         $member = Member::get()->byID($memberID);
 
         // Continue if we have a valid member
         if ($member && $member instanceof Member) {
-            if (!isset($data['token'])) {
+            if (!$token) {
                 $member->registerFailedLogin();
 
                 $result->addError(_t(self::class . '.NOTOKEN', 'No token sent'));
             } else {
-                $secret = Base32::decode($member->TOTPSecret);
-                $algorithm = self::get_algorithm();
-                $totp = new Totp($algorithm);
-                $key = $totp->GenerateToken($secret);
-                $user_submitted_key = $data['token'];
+                /** @var TOTPProvider $provider */
+                $provider = Injector::inst()->get(TOTPProvider::class);
```

Comment:
I wonder if this might be better as a dependency?

```php
private static $dependencies = [
    'totpProvider' => TOTPProvider::class,
];

protected $totpProvider;

public function setTOTPProvider(TOTPProvider $provider) ;
public function getTOTPProvider();
```